Prepare RemoteLayerTreeDrawingArea to draw with multiple root frames

achristensen07 opened the pull request:
#### c36b5e9890fea75c57b7d533a4b05d8bb5a791f9
<pre>
Prepare RemoteLayerTreeDrawingArea to draw with multiple root frames
<a href="https://bugs.webkit.org/show_bug.cgi?id=256396">https://bugs.webkit.org/show_bug.cgi?id=256396</a>
rdar://108965687

Reviewed by Simon Fraser.

In order to support drawing a page with site isolation enabled and two iframes
that have loaded the same domain, RemoteLayerTreeDrawingArea needs to not have
any special behavior for its first root layer.  In order to do this, I need to
not add the first root layer in the RemoteLayerTreeDrawingArea constructor, but
rather add each root layer with a call to addRootLayer.  I did add a few places
where we iterate a Vector checking each member, which isn&apos;t ideal, but in the
default non-site-isolated configuration that vector will always contain 1 element,
so it&apos;s still O(1).  As we progress with the site isolation project, we may want
to switch to a HashMap, but having a WebFrame* as a key is a little strange.
In the meantime, this is just a step in the right direction.

* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.cpp:
(WebKit::WebChromeClient::contentsSizeChanged const):
(WebKit::WebChromeClient::attachViewOverlayGraphicsLayer):
* Source/WebKit/WebProcess/WebPage/DrawingArea.h:
(WebKit::DrawingArea::mainFrameContentSizeChanged):
(WebKit::DrawingArea::addRootFrame):
(WebKit::DrawingArea::attachViewOverlayGraphicsLayer):
(WebKit::DrawingArea::attachToInitialRootFrame): Deleted.
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/RemoteLayerTreeDrawingArea.h:
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/RemoteLayerTreeDrawingArea.mm:
(WebKit::RemoteLayerTreeDrawingArea::RemoteLayerTreeDrawingArea):
(WebKit::RemoteLayerTreeDrawingArea::attachViewOverlayGraphicsLayer):
(WebKit::RemoteLayerTreeDrawingArea::addRootFrame):
(WebKit::RemoteLayerTreeDrawingArea::mainFrameContentSizeChanged):
(WebKit::initialRootFrame): Deleted.
(WebKit::m_updateRenderingTimer): Deleted.
(WebKit::RemoteLayerTreeDrawingArea::attachToInitialRootFrame): Deleted.
* Source/WebKit/WebProcess/WebPage/WebFrame.cpp:
(WebKit::WebFrame::transitionToLocal):
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::m_appHighlightsVisible):
(WebKit::WebPage::reinitializeWebPage):
* Source/WebKit/WebProcess/WebPage/mac/RemoteLayerTreeDrawingAreaMac.h:
* Source/WebKit/WebProcess/WebPage/mac/RemoteLayerTreeDrawingAreaMac.mm:
(WebKit::RemoteLayerTreeDrawingAreaMac::mainFrameContentSizeChanged):
* Source/WebKit/WebProcess/WebPage/mac/TiledCoreAnimationDrawingArea.h:
* Source/WebKit/WebProcess/WebPage/mac/TiledCoreAnimationDrawingArea.mm:
(WebKit::TiledCoreAnimationDrawingArea::attachViewOverlayGraphicsLayer):
(WebKit::TiledCoreAnimationDrawingArea::mainFrameContentSizeChanged):

Canonical link: <a href="https://commits.webkit.org/263746@main">https://commits.webkit.org/263746@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/327f6b4a24db5507ade6b50948d301f5430d88a9

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/5670 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/5826 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/6016 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/7222 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/5660 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/6048 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/5798 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/7820 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/5776 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/15/builds/5823 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/5081 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/7271 "Built successfully") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/39/builds/3300 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/12732 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/5168 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/5182 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/7018 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/5619 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/4597 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/5066 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/9180 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/647 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/5427 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebKit/WebProcess/WebPage/DrawingArea.h:
```diff
@@ -138,7 +138,7 @@ class DrawingArea : public IPC::MessageReceiver, public WebCore::DisplayRefreshM
     virtual void tryMarkLayersVolatile(CompletionHandler<void(bool)>&&);
 
-    virtual void attachViewOverlayGraphicsLayer(WebCore::GraphicsLayer*) { }
+    virtual void attachViewOverlayGraphicsLayer(WebFrame&, WebCore::GraphicsLayer*) { }
```
Why does this take a `WebFrame` and not a `WebCore::FrameIdentifier`?

smfr commented on Source/WebKit/WebProcess/WebPage/RemoteLayerTree/RemoteLayerTreeDrawingArea.h:
```diff
@@ -109,7 +109,7 @@ class RemoteLayerTreeDrawingArea : public DrawingArea, public WebCore::GraphicsL
     void setDeviceScaleFactor(float) override;
 
-    void mainFrameContentSizeChanged(const WebCore::IntSize&) override;
+    void mainFrameContentSizeChanged(WebCore::FrameIdentifier, const WebCore::IntSize&) override;
```
main or root frame?

smfr commented on Source/WebKit/WebProcess/WebPage/RemoteLayerTree/RemoteLayerTreeDrawingArea.h:
```diff
@@ -71,9 +71,9 @@ class RemoteLayerTreeDrawingArea : public DrawingArea, public WebCore::GraphicsL
     WebCore::GraphicsLayerFactory* graphicsLayerFactory() override;
     void setRootCompositingLayer(WebCore::Frame&, WebCore::GraphicsLayer*) override;
-    void attachToInitialRootFrame(WebCore::FrameIdentifier) final;
+    void addRootFrame(WebCore::FrameIdentifier) final;
```
Don't we need a corresponding `removeRootFrame` ?

smfr commented on Source/WebKit/WebProcess/WebPage/RemoteLayerTree/RemoteLayerTreeDrawingArea.mm:
```diff
@@ -147,17 +136,31 @@
+    ASSERT(webFrame);
+
+    auto layer = GraphicsLayer::create(graphicsLayerFactory(), *this);
+    layer->setName(MAKE_STATIC_STRING_IMPL("drawing area root"));
```
Would be good to put the frameID into the name

smfr commented on Source/WebKit/WebProcess/WebPage/RemoteLayerTree/RemoteLayerTreeDrawingArea.mm:
```diff
@@ -147,17 +136,31 @@
+        WTFMove(layer),
+        nullptr,
+        nullptr,
+        webFrame
```
Might be nicer to store FrameIdentifiers in this array.

smfr commented on Source/WebKit/WebProcess/WebPage/RemoteLayerTree/RemoteLayerTreeDrawingArea.mm:
```diff
@@ -459,10 +462,17 @@
-    m_rootLayers[0].layer->setSize(contentsSize);
+    auto* webFrame = WebProcess::singleton().webFrame(frameID);
+    ASSERT(webFrame);
+    for (auto& rootLayerInfo : m_rootLayers) {
```
Rather than the loop with confusing continue/return, using find() would be nicer.

smfr commented on Source/WebKit/WebProcess/WebPage/WebPage.cpp:
```diff
@@ -763,6 +763,9 @@ WebPage::WebPage(PageIdentifier pageID, WebPageCreationParameters&& parameters)
 
     m_drawingArea = DrawingArea::create(*this, parameters);
+    // FIXME: Rename coreAbstractFrame to coreFrame and coreFrame to coreLocalFrame.
+    if (!receivedMainFrameIdentifierFromUIProcess)
```
`receivedMainFrameIdentifierFromUIProcess` seems a bit magic.

achristensen07 committed:
Prepare RemoteLayerTreeDrawingArea to draw with multiple root frames
https://bugs.webkit.org/show_bug.cgi?id=256396
rdar://108965687

Reviewed by Simon Fraser.

In order to support drawing a page with site isolation enabled and two iframes
that have loaded the same domain, RemoteLayerTreeDrawingArea needs to not have
any special behavior for its first root layer.  In order to do this, I need to
not add the first root layer in the RemoteLayerTreeDrawingArea constructor, but
rather add each root layer with a call to addRootLayer.  I did add a few places
where we iterate a Vector checking each member, which isn't ideal, but in the
default non-site-isolated configuration that vector will always contain 1 element,
so it's still O(1).  As we progress with the site isolation project, we may want
to switch to a HashMap, but having a WebFrame* as a key is a little strange.
In the meantime, this is just a step in the right direction.

* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.cpp:
(WebKit::WebChromeClient::contentsSizeChanged const):
(WebKit::WebChromeClient::attachViewOverlayGraphicsLayer):
* Source/WebKit/WebProcess/WebPage/DrawingArea.h:
(WebKit::DrawingArea::mainFrameContentSizeChanged):
(WebKit::DrawingArea::addRootFrame):
(WebKit::DrawingArea::attachViewOverlayGraphicsLayer):
(WebKit::DrawingArea::attachToInitialRootFrame): Deleted.
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/RemoteLayerTreeDrawingArea.h:
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/RemoteLayerTreeDrawingArea.mm:
(WebKit::RemoteLayerTreeDrawingArea::RemoteLayerTreeDrawingArea):
(WebKit::RemoteLayerTreeDrawingArea::attachViewOverlayGraphicsLayer):
(WebKit::RemoteLayerTreeDrawingArea::addRootFrame):
(WebKit::RemoteLayerTreeDrawingArea::mainFrameContentSizeChanged):
(WebKit::initialRootFrame): Deleted.
(WebKit::m_updateRenderingTimer): Deleted.
(WebKit::RemoteLayerTreeDrawingArea::attachToInitialRootFrame): Deleted.
* Source/WebKit/WebProcess/WebPage/WebFrame.cpp:
(WebKit::WebFrame::transitionToLocal):
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::m_appHighlightsVisible):
(WebKit::WebPage::reinitializeWebPage):
* Source/WebKit/WebProcess/WebPage/mac/RemoteLayerTreeDrawingAreaMac.h:
* Source/WebKit/WebProcess/WebPage/mac/RemoteLayerTreeDrawingAreaMac.mm:
(WebKit::RemoteLayerTreeDrawingAreaMac::mainFrameContentSizeChanged):
* Source/WebKit/WebProcess/WebPage/mac/TiledCoreAnimationDrawingArea.h:
* Source/WebKit/WebProcess/WebPage/mac/TiledCoreAnimationDrawingArea.mm:
(WebKit::TiledCoreAnimationDrawingArea::attachViewOverlayGraphicsLayer):
(WebKit::TiledCoreAnimationDrawingArea::mainFrameContentSizeChanged):

Canonical link: https://commits.webkit.org/263746@main
